blackbird_python/blackbird/utils.py

```diff
@@ -0,0 +1,245 @@
+            if q not in grid:
+                # add a new line to the circuit
+                grid[q] = []
+
+            grid[q].append([idx, cmd])
+
+    G = nx.DiGraph()
+
+    for q, cmds in grid.items():
+        if cmds:
+            # add the first operation on the wire that does not depend on anything
+            attrs = cmds[0][1]._asdict()
+            G.add_node(cmds[0][0], **attrs)
+
+        for i in range(1, len(cmds)):
+            # add the edge between the operations, and the operation nodes themselves
+            if cmds[i][0] not in G:
+                attrs = cmds[i][1]._asdict()
+                G.add_node(cmds[i][0], **attrs)
+
+            G.add_edge(cmds[i-1][0], cmds[i][0])
+
+    return G
+
+
+def match_template(template, program):
+    """Match a template against a Blackbird program, returning template
+    parameter values.
+
+    For example, consider the following template and program:
+
+    .. code-block:: python
+
+        template = blackbird.loads(\"\"\"\\
+        name prog
+        version 1.0
+
+        Dgate(-{r}, 0.45) | 1
+        Vac | 2
+        Sgate({r}, 2*{phi}-1) | 0
+        \"\"\")
+
+        program = blackbird.loads(\"\"\"\\
+        name prog
+        version 1.0
+
+        Sgate(0.543, -1.432*pi) | 0
+        Dgate(-0.543, 0.45) | 1
+        Vac | 2
+        \"\"\")
+
+    By applying the ``match_template`` function, we can match the template parameters:
+
+    >>> res = match_template(template, program)
+    >>> print(res)
+    {'r': 0.543, 'phi': -1.74938033997029}
+
+    Verifying this is correct:
+
+    >>> print((-1.432*np.pi+1)/2)
+    -1.7493803399702919
+
+    .. note::
+
+        The template and the Blackbird program to match must have the *same*
+        version number and target, otherwise an :class:`TemplateError` will be raised.
+
+    Args:
+        template (blackbird.Program): the Blackbird template
+        program (blackbird.Program): a Blackbird program to match against
+            the template
+
+    Returns:
+        dict[str, Number]: mapping from the template parameter name
+        to a numerical value.
+    """
+
+    # check template
+    if not template.is_template():
+        raise TemplateError("Argument 1 is not a template!")
+
+    if program.is_template():
+        raise TemplateError("Argument 2 cannot be a template!")
+
+    if template.version != program.version:
+        raise TemplateError("Mismatching Blackbird version between template and program")
+
+    if template.target['name'] != program.target['name']:
+        raise TemplateError("Mismatching target between template and program")
+
+    G1 = to_DiGraph(template)
+    G2 = to_DiGraph(program)
+
+    def node_match(n1, n2):
+        """Returns True if both nodes have the same name"""
+        return n1['name'] == n2['name'] and n1['modes'] == n2['modes']
+
+    GM = isomorphism.DiGraphMatcher(G1, G2, node_match)
+
+    # check if topology matches
+    if not GM.is_isomorphic():

```

Comment:
how does this algorithm scale?

Reply:
Here is the algorithm used: http://citeseerx.ist.psu.edu/viewdoc/summary?doi=10.1.1.101.5342

Not polynomial, but fast enough in practise for the circuit sizes we will be considering in the near term.